velox/exec/tests/MultiFragmentTest.cpp

```diff
@@ -1780,5 +1780,49 @@ DEBUG_ONLY_TEST_F(MultiFragmentTest, exchangeStatsOnFailure) {
   ASSERT_TRUE(waitForTaskCompletion(producerTask.get(), 3'000'000));
 }
 
+TEST_F(MultiFragmentTest, memoryAbortTest) {
+  setupSources(1, 10);
+  std::vector<std::shared_ptr<TempFilePath>> filePaths(
+      filePaths_.begin(), filePaths_.end());
+
+  auto partialSortTaskId = makeTaskId("partialSortBy", 0);
+  auto planNodeIdGenerator = std::make_shared<core::PlanNodeIdGenerator>();
+  auto partialSortPlan = PlanBuilder(planNodeIdGenerator)
+                             .localMerge(
+                                 {"c0"},
+                                 {PlanBuilder(planNodeIdGenerator)
+                                      .tableScan(rowType_)
+                                      .orderBy({"c0"}, true)
+                                      .planNode()})
+                             .partitionedOutput({}, 1)
+                             .planNode();
+
+  auto partialSortTask = makeTask(partialSortTaskId, partialSortPlan, 1);
+  Task::start(partialSortTask, 1);
+  addHiveSplits(partialSortTask, filePaths);
+  RowTypePtr outputType = partialSortPlan->outputType();
+
+  auto finalSortTaskId = makeTaskId("finalSortBy", 0);
+  auto finalSortPlan = PlanBuilder()
+                           .mergeExchange(outputType, {"c0"})
+                           .partitionedOutput({}, 1)
+                           .planNode();
+
+  auto finalSortTask = makeTask(finalSortTaskId, finalSortPlan, 0);
+  try {
+    VELOX_FAIL("memoryAbortTest");
+  } catch (const VeloxRuntimeError& e) {
+    finalSortTask->pool()->abort(std::current_exception());
```

Comment:
Which code path in Prestissimo ends up calling task->pool()->abort without calling task->requestAbort first?

Reply:
This is triggered by memory arbitration internally when memory arbitration fails to allocate memory and has to abort one. I just run into this when I add memory pool abort test (which is supposed not happening in production) and found the task hanging problem. And realized that similar race could happen in Prestissimo with the concurrent updates from coordinator.

Reply:
> And realized that similar race could happen in Prestissimo with the concurrent updates from coordinator.

I'm not sure I understand. Which API call from coordinator would end up calling pool->abort and not task->requestAbort or some other Task API?